### ISSUE TITLE
reStructuredText documents with markdown extension in doc/

### ISSUE THREAD

**Issue by nirs** [nim-lang/Nim#20928]
### Description

Most of the *.md files in doc/ are actually reStructuredText documents so they render incorrectly in github or your editor.

Candidates:
```
$ git grep -l -E '^\.\. ' *.md
apis.md
backends.md
contributing.md
destructors.md
docgen.md
drnim.md
estp.md
filters.md
hcr.md
idetools.md
intern.md
koch.md
lib.md
manual.md
manual_experimental.md
manual_experimental_strictnotnil.md
markdown_rst.md
mm.md
nep1.md
nimc.md
nimfix.md
nimgrep.md
niminst.md
nims.md
nimsuggest.md
overview.md
testament.md
tools.md
tut1.md
tut2.md
tut3.md
```

Renaming to *.rst fixes some of the files, but some files have syntax errors, since they have
mixed reStructuredText and markdown syntax:

```
$ rst2html backends.md 
backends.md:75: (ERROR/3) Unknown interpreted text role "idx".
backends.md:97: (ERROR/3) Unknown interpreted text role "idx".
backends.md:166: (ERROR/3) Unexpected indentation.
backends.md:163: (WARNING/2) Inline literal start-string without end-string.
backends.md:163: (WARNING/2) Inline interpreted text or phrase reference start-string without end-string.
backends.md:167: (WARNING/2) Block quote ends without a blank line; unexpected unindent.
backends.md:168: (SEVERE/4) Unexpected section title.
...
Exiting due to level-4 (SEVERE) system message.
```

### Nim Version

Devel

### Possible Solution

Looks like the quick way to fix is to rename the files and replace the markdown sections
with rst syntax.

**Comment by nirs** [comment]
Wrong rendering as markdown:
https://github.com/nirs/Nim/blob/devel/doc/apis.md

Fixed by renaming:
https://github.com/nirs/Nim/blob/doc-rst/doc/apis.rst

Wrong rendering as markdown:
https://github.com/nirs/Nim/blob/devel/doc/destructors.md

Also wrong as rst, mixing reStructuredText and markdown:
https://github.com/nirs/Nim/blob/doc-rst/doc/destructors.rst

In older version we have correctly rendered rst files:
https://github.com/nim-lang/Nim/blob/version-1-6/doc/tut1.rst

Looks this is relevant commit:
417b90a7e5b88bfc0ad1bfbbc81a3205c99e128e

@a-mr is this incomplete conversion to markdown?

**Comment by a-mr** [comment]
1) The only major problem I see is with tables.

   There is a popular Github Markdown syntax extension for tables also implemented in Nim (though tables are not supported in original Markdown nor in CommonMark spec).

   But it seems ugly for large tables  because it does not support multi-line rows in input files. E.g. if we have this row in RST:
   ```rst
   delete                  delete, del    del is supposed to be faster than
                                          delete, because it does not keep
                                          the order; delete keeps the order
   ```

   In Markdown that would be:

   ```markdown
   delete        |          delete, del  |  del is supposed to be faster than delete, because it       does not keep the order; delete keeps the order
   ```

   That's why I hesitated to start migration to Markdown here.

   Pandoc Markdown has a very different support for tables: https://pandoc.org/MANUAL.html#extension-simple_tables , which closely resembles RST simple tables (with substition `-` -> `=`). May be it's better to pretend that we want to use Pandoc Markdown extension instead of Github one?

2) the fact that `:idx:` role is not known to RST and Markdown is OK, it should not do any harm. There is a Pandoc Markdown syntax `{.idx}` for that purpose but AFAIK Github Markdown would not support it anyway.

3) file header with fields like "author", etc, is not supported in Markdown. I plan to implement Pandoc [yaml metadata block](https://pandoc.org/MANUAL.html#extension-yaml_metadata_block) extension for that and other purposes. Currently it's just ugly-rendered string in the beginning of file, no harm really.

@nirs What else have you spotted?

**Comment by nirs** [comment]
I just started to look at the docs, so I don't have a complete picture.

version-1.6 rst looks good at first look:
https://github.com/nim-lang/Nim/blob/version-1-6/doc/tut1.rst

But looking the raw source the code examples are missing:
https://github.com/nim-lang/Nim/blob/version-1-6/doc/tut1.rst#for-statement

```rst
For statement
-------------

The `for` statement is a construct to loop over any element an *iterator*
provides. The example uses the built-in `countup
<system.html#countup.i,T,T,Positive>`_ iterator:

.. code-block:: nim
    :test: "nim c $1"
  echo "Counting to ten: "
  for i in countup(1, 10):
    echo i
  # --> Outputs 1 2 3 4 5 6 7 8 9 10 on different lines
```

This can be fixed by leaving blank line after the code-block:
https://github.com/nirs/Nim/blob/tut1-rst/doc/tut1.rst#for-statement

Maybe it possible to add missing features to markdown, but will it work
in github? Is this going to be easier for contributors to use Nim
special markdown flavor?

**Comment by a-mr** [comment]
> This can be fixed by leaving blank line after the code-block

Yes, it's a known problem and I proposed a fix: https://github.com/nim-lang/Nim/pull/19878 but since that very (refused) PR things took a different turn: Araq started to insist that we move to Markdown completely, and even preferably drop RST support!

Personally I am an advocate for moving towards support for both Markdown and RST in our parser but separate them as much as possible using [recently introduced doctype pragma](https://github.com/nim-lang/Nim/pull/20252). Personally I prefer RST actually, it's much more well-devised than Markdown.

> Maybe it possible to add missing features to markdown, but will it work
in github?

That's the problem that it won't. Actually RST features can be seen as an extension to Markown. And current state of things is that most RST features are still enabled in (default) Markdown mode. Though I'm planning to disable them gradually.

In many cases RST syntax does not cause any problems.

> Is this going to be easier for contributors to use Nim
special markdown flavor?

Certainly not. But if we really want to move towards Markdown-ish markup as default option and not throw away any features, then that's the only way. I'm just moving opportunistically here -- I try to find compromise between my preferences and Araq's desires. There is no solution that would satisfy everyone.

**Comment by elcritch** [comment]
Are you using pandoc to do the conversion? I believe there's options to use various markdown formats. If you're doing it by hand it might be a big improvement to use pandoc with the appropriate markdown target.  

Here's what looks to be the github markdown format: https://pandoc.org/MANUAL.html#markdown-variants and looks like some stackoverflow notes on it: https://stackoverflow.com/questions/66586079/customizing-gfm-markdown-output-in-pandoc
